### PR TITLE
extension(tests): Add extension smoketest

### PR DESCRIPTION
**Pull request by wardpeet**
This PR will run the extension in chrome and checks if the report has all necessary fields listed.

the manifest change is needed because puppeteer can't detect the active tab. We need to add tabs to the permission list, we could temporary add it for each run so we don't have to update our manifest.

Also a caveat, we need to run puppeteer in non headless mode to let the extensions work

### CODE REVIEW

**Comment by wardpeet** [comment]
Fixes #4479

File: lighthouse-extension/test/extension.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                awaitPromise: true,
+                returnByValue: true,
+              }
+            );
+          })
+          .then(() => {
+            return browser.pages().then(pages => {
+              const page = pages.find(page => page.url().includes('blob:chrome-extension://'));
+              const assertAudits = (category, expected, selector) =>
+                page.$eval(`#${category}`,
```
Comment (paulirish) [inline lighthouse-extension/test/extension.js]: trying this out to see if it's more clear:

```js
function assertAuditElements(category, expected, selector) {
  page.$evaluate(({category, selector}) => 
    document.querySelector(`#${category}`).parentNode.querySelectorAll(selector).length,
    {category, selector}
  ).then(elementCount => {
    assert.equal(expected, elementCount);
  });
}

```

File: lighthouse-extension/test/extension.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      return page.goto('https://www.paulirish.com', {waitUntil: 'networkidle2'});
+    })
+    .then(() => browser.targets())
+    .then(targets => {
+      const extensionTarget = targets.find(target => {
+        return target._targetInfo.title === 'Lighthouse';
+      });
+
+      if (extensionTarget) {
+        extensionTarget.createCDPSession()
```
Comment (ebidel) [inline lighthouse-extension/test/extension.js]: Any way to use puppeteer's `evaluate` method(s) here 

https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#executioncontextevaluatepagefunction-args maybe?

```js
await ctx.evaluate(() => {
  return runLighthouseInExtension({restoreCleanState: true}, ['performance', 'pwa', 'accessibility', 'best-practices', 'seo']);
});
```

Whatever that looks like. Not sure if you can get the crx context, other than what you're doing using an private api.

File: lighthouse-extension/test/extension.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+const config = require(path.resolve(__dirname, '../../lighthouse-core/config/default.js'));
+
+puppeteer.launch({
+  headless: false,
+  executablePath: process.env.CHROME_PATH ? process.env.CHROME_PATH : undefined,
+  args: [
+    `--disable-extensions-except=${lighthouseExtensionPath}`,
+    `--load-extension=${lighthouseExtensionPath}`,
+  ],
+}).then(browser => {
```
Comment (paulirish) [inline lighthouse-extension/test/extension.js]: Let's do async/await for this file. I'm sorry we didn't sort that out beforehand.

But IMO that'll make this guy a lot more readable.

File: lighthouse-extension/test/extension.js (hunk 1 above)
Comment (ebidel) [inline lighthouse-extension/test/extension.js]: Having a hard time parsing this. Maybe just use `function`. And I think selector can just be passed as a string, not in in an obj:

```js
const assertNumAuditsEqual  = function(el, selector) {
  return page.$eval(`#${category}`, (el, selector) => el.parentNode.querySelectorAll(selector).length, selector));
}
```

Comment (paulirish) [reply lighthouse-extension/test/extension.js]: also, let's use a keyed object for the method signature?   


```js
assertAudits({
	category: 'performance',
	count: config.categories.performance.audits.length,
	selector: '.lh-audit,.lh-timeline-metric,.lh-perf-hint,.lh-filmstrip'
});
```

File: lighthouse-extension/app/manifest.json [hunk 4]
```diff
@@ -18,6 +18,7 @@
   },
   "permissions": [
     "activeTab",
+    "tabs",
```
Comment (paulirish) [inline lighthouse-extension/app/manifest.json]: i remember you telling me, but remind me again why we now need this permission?

File: .travis.yml [hunk 5]
```diff
@@ -19,24 +19,24 @@ cache:
...
-  - echo $TRAVIS_EVENT_TYPE;
-  - echo $TRAVIS_BRANCH
   - yarn bundlesize
   - yarn lint
   - yarn unit
   - yarn type-check
   - yarn closure
   - yarn smoke
   - yarn smokehouse
+  - node lighthouse-extension/test/extension.js
```
Comment (paulirish) [inline .travis.yml]: let's add an npm scripts entry to the extensions package.json so it's easier to run on demand

    cd lighthouse-extension; yarn test:e2e

File: lighthouse-extension/test/extension.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                .then(categories => {
+                  assert.equal(categories.length, 5);
+
+                  return assertAudits(
+                    'performance',
+                    config.categories.performance.audits.length,
+                    '.lh-audit,.lh-timeline-metric,.lh-perf-hint,.lh-filmstrip'
+                  );
+                })
+                .then(
```
Comment (ebidel) [inline lighthouse-extension/test/extension.js]: the formatting of these then's should be the same. I may also break this testing section into a separate function if you can.

File: lighthouse-extension/app/manifest.json (hunk 4 above)
Comment (patrickhulce) [reply lighthouse-extension/app/manifest.json]: let's definitely add this on demand as you suggested @wardpeet, grabbing a new permission might still disable the extension on update?

Comment (ebidel) [reply lighthouse-extension/app/manifest.json]: I don't think we need it.  In https://stackoverflow.com/questions/48089670/detect-and-test-chrome-extension-using-puppeteer, I needed it to get the title of tab. We're not doing anything special with the tabs api here.

[the remark this answers is not in the document]
Comment (wardpeet) [reply lighthouse-extension/app/manifest.json]: I could change the config just during the test so we don't need to change it for installations?

File: .travis.yml [hunk 7]
```diff
@@ -19,24 +19,24 @@ cache:
...
     - /home/travis/.rvm/gems/
 install:
+  - export PUPPETEER_SKIP_CHROMIUM_DOWNLOAD=1
   - yarn
   # travis can't handle the parallel install (without caches)
   - yarn run install-all:task:windows
 before_script:
   - export DISPLAY=:99.0
-  - export CHROME_PATH="$(pwd)/chrome-linux/chrome"
+  - export CHROME_PATH="$(which google-chrome-stable)"
```
[the remark this answers is not in the document]
Comment (wardpeet) [reply .travis.yml]: I just thought it was a bit stupid to install it twice and we moved to chrome stable to test the lowest version we support. I could enable chromium again.

File: lighthouse-extension/test/extension.js [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+const path = require('path');
+const assert = require('assert');
+const puppeteer = require('puppeteer');
+
+const lighthouseExtensionPath = path.resolve(__dirname, '../app');
+const config = require(path.resolve(__dirname, '../../lighthouse-core/config/default.js'));
+
+puppeteer.launch({
+  headless: false,
+  executablePath: process.env.CHROME_PATH ? process.env.CHROME_PATH : undefined,
```
[the remark this answers is not in the document]
Comment (wardpeet) [reply lighthouse-extension/test/extension.js]: Thanks!

File: lighthouse-extension/app/manifest.json (hunk 4 above)
Comment (wardpeet) [reply lighthouse-extension/app/manifest.json]: The reason why I need the extension is the following function
```
 const queryOpts = {
        active: true,
        currentWindow: true,
      };

      chrome.tabs.query(queryOpts, (tabs => {
```
without it I couldn't get the url of the tab

File: lighthouse-extension/test/extension.js (hunk 3 above)
Comment (wardpeet) [reply lighthouse-extension/test/extension.js]: what about node 6 which we run our PR checks on? it doesn't support async/await right?

File: lighthouse-extension/test/extension.js (hunk 2 above)
Comment (wardpeet) [reply lighthouse-extension/test/extension.js]: sounds neat, will try :)

File: .travis.yml (hunk 7 above)
[the remark this answers is not in the document]
Comment (paulirish) [reply .travis.yml]: well, it does _appear_ to work fine with chrome stable. 
I'm also fine with using the bundled Chromium, but yes it does seem wasteful to d/l each time. 

If it works fine with chrome stable as of now, then moving forward (assuming we infrequently bump the pptr version), it shouldn't regress. So I'm +1 to using the stable.

File: lighthouse-extension/app/manifest.json (hunk 4 above)
Comment (paulirish) [reply lighthouse-extension/app/manifest.json]: Yessssss. I repro'd this.
Basically I believe activeTab isn't enough because of however we're bypassing the popup and such. It's hitting some corner case where `activeTab` just isn't enough. +1 to dynamically adding tabs permission just for this case.

File: lighthouse-extension/test/extension.js (hunk 3 above)
Comment (paulirish) [reply lighthouse-extension/test/extension.js]: we're about to drop node 6 support. :) we'll land this patch in the 3.0 side of things.

File: lighthouse-extension/test/extension.js (hunk 2 above)
Comment (paulirish) [reply lighthouse-extension/test/extension.js]: yup +1 to using this rather than dropping to the protocol. good call.

Comment (wardpeet) [reply lighthouse-extension/test/extension.js]: @ebidel not sure how I can get that execution context from the Target class. I tried `.page().mainFrame().executionContext()` but page() is of course null.

Any ideas? 😛

File: lighthouse-extension/package-lock.json [hunk 9]
```diff
@@ -0,0 +1,5979 @@
+{
```
Comment (paulirish) [inline lighthouse-extension/package-lock.json]: we're still yarn folk so we should be updating the yarn lock and not adding this, right?
or did you already point out why we need this and i'm forgetting?

File: lighthouse-extension/test/extension.js [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+  const page = await browser.newPage();
+  await page.goto('https://www.paulirish.com', {waitUntil: 'networkidle2'});
+  const targets = await browser.targets();
+  const extensionTarget = targets.find(({_targetInfo}) => {
+    return _targetInfo.title === 'Lighthouse' &&
+      _targetInfo.type === 'background_page';
+  });
+
+  if (extensionTarget) {
```
Comment (paulirish) [inline lighthouse-extension/test/extension.js]: early exit instead

File: lighthouse-extension/test/extension.js [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+const path = require('path');
+const assert = require('assert');
+const fs = require('fs');
+const puppeteer = require('puppeteer');
+
+const lighthouseExtensionPath = path.resolve(__dirname, '../app');
+const config = require(path.resolve(__dirname, '../../lighthouse-core/config/default.js'));
+const lighthouseCategories = Object.keys(config.categories);
+
+const manifestLocation = path.join(lighthouseExtensionPath, 'manifest.json');
```
Comment (paulirish) [inline lighthouse-extension/test/extension.js]: can you pull this stuff into a little named function? might as well create another for putting the orig back.

Comment (paulirish) [reply lighthouse-extension/test/extension.js]: actually if we use mocha to run this test, then we can use before/after methods for these things.

seems fairly reasonable to use mocha at this point

File: lighthouse-extension/test/extension.js [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+          throw new Error(lighthouseResult.exceptionDetails.exception.description);
+        }
+
+        throw new Error(lighthouseResult.exceptionDetails.text);
+      }
+
+      extensionPage = (await browser.pages())
+        .find(page => page.url().includes('blob:chrome-extension://'));
+
+      const categories = await extensionPage.$$(`#${lighthouseCategories.join(',#')}`);
```
Comment (paulirish) [inline lighthouse-extension/test/extension.js]: i'd pull 80-94 into a separate function, so it's clear that this is the test assertions against the report page

File: lighthouse-extension/package-lock.json (hunk 9 above)
Comment (wardpeet) [reply lighthouse-extension/package-lock.json]: no no, it's my mistake . It shouldn't be here. I use yarn in WSL but npm on powershell because i'm to lazy to install yarn on windows as I only need it to run puppeteer (i'll probably create a bug for that in the puppeteer repository).

File: lighthouse-extension/test/extension.js (hunk 11 above)
Comment (wardpeet) [reply lighthouse-extension/test/extension.js]: any idea how I should name this test? describe and it text😛.

File: lighthouse-extension/test/extension.js (hunk 2 above)
Comment (ebidel) [reply lighthouse-extension/test/extension.js]: Asking...

File: lighthouse-extension/package-lock.json (hunk 9 above)
Comment (paulirish) [reply lighthouse-extension/package-lock.json]: fwiw `npm install -g yarn` is what we do on appveyor. probably not too bad.

Comment (paulirish) [reply lighthouse-extension/package-lock.json]: also i think pptr is moving to package-lock soon.

File: lighthouse-extension/test/extension.js (hunk 11 above)
Comment (paulirish) [reply lighthouse-extension/test/extension.js]: describe('lighthouse chrome extension', ...
it('completes an end-to-end run', ...

File: lighthouse-extension/package.json [hunk 13]
```diff
@@ -22,6 +23,7 @@
     "gulp-util": "^3.0.7",
     "gulp-zip": "^3.2.0",
     "package-json-versionify": "^1.0.4",
+    "puppeteer": "^1.1.1",
```
Comment (brendankenny) [inline lighthouse-extension/package.json]: what do folks think about moving puppeteer up to a root-level dev dependency? As I was testing viewer yesterday I was thinking it would be great to have a similar (but should be much simpler) e2e test for viewer, and it seems a bit silly to have a copy here and in viewer when `yarn install-all` would install both anyway...

Comment (ebidel) [reply lighthouse-extension/package.json]: Great idea. Viewer UI testing is the perfect use case for 🤹‍♀️

Comment (paulirish) [reply lighthouse-extension/package.json]: sgtm

**Comment by wardpeet** [comment]
@paulirish yarn compile-devtools is failing. How can I fix that?

**Comment by paulirish** [comment]
>  yarn compile-devtools is failing. How can I fix that?

You can't. :) I filed #4735 to look into it. I'll sort it out with brendan.

File: lighthouse-extension/test/extension-test.js [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+  assert.equal(categories.length, lighthouseCategories.length,
+    `${categories.join(' ')} does not match ${lighthouseCategories.join(' ')}`);
+
+  for (const category of lighthouseCategories) {
+    let selector = '.lh-audit';
+    let expected = config.categories[category].audits.length;
+    if (category === 'performance') {
+      selector = '.lh-audit,.lh-timeline-metric,.lh-perf-hint,.lh-filmstrip';
+      // we deducted the expected by 1 because network-requests audit is not showing in the report
+      expected = expected - 1;
```
Comment (paulirish) [inline lighthouse-extension/test/extension-test.js]: i think what we want here is more like

```js
expected = config.categories[category].audits.map(a => !!a.group).length;
```

File: lighthouse-core/scripts/run-mocha.sh [hunk 15]
```diff
@@ -18,10 +18,13 @@ elif [ "$flag" == '--cli' ]; then
   _runmocha 'lighthouse-cli'
 elif [ "$flag" == '--viewer' ]; then
   _runmocha 'lighthouse-viewer'
+elif [ "$flag" == '--extension' ]; then
```
Comment (brendankenny) [inline lighthouse-core/scripts/run-mocha.sh]: I missed the earlier conversation, but do we want this in the unit tests? Considerably slower than (most? all?) the rest of them. It's also not really a unit test...

`yarn test-extension` or something instead?

Comment (brendankenny) [reply lighthouse-core/scripts/run-mocha.sh]: having it under `test` in the extension `package.json` would make sense as well

File: lighthouse-extension/test/extension-test.js [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+const fs = require('fs');
+const puppeteer = require('puppeteer');
+
+const lighthouseExtensionPath = path.resolve(__dirname, '../app');
+const config = require(path.resolve(__dirname, '../../lighthouse-core/config/default.js'));
+const lighthouseCategories = Object.keys(config.categories);
+
+let extensionPage;
+
+const assertAuditElements = async ({category, expected, selector}) => {
```
Comment (brendankenny) [inline lighthouse-extension/test/extension-test.js]: nit: any reason to have these two as arrow functions expressions? e.g. could just be
`async function assertAuditElements({category, expected, selector}) {`

File: lighthouse-extension/test/extension-test.js [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+          throw new Error(lighthouseResult.exceptionDetails.exception.description);
+        }
+
+        throw new Error(lighthouseResult.exceptionDetails.text);
+      }
+
+      extensionPage = (await browser.pages())
+        .find(page => page.url().includes('blob:chrome-extension://'));
+    } catch (err) {
+      assert.ok(false, err.message);
```
Comment (brendankenny) [inline lighthouse-extension/test/extension-test.js]: any reason for the try/catch in this case if it's always going to fail on this? (shouldn't mocha treat the exception as a test failure, with message `err.message` anyways?)

File: lighthouse-extension/test/extension-test.js [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    if (extensionPage) {
+      await assertReport();
+    }
+
+    if (browser) {
+      await browser.close();
```
Comment (brendankenny) [inline lighthouse-extension/test/extension-test.js]: if the test throws, do we want this pulled out into `after` so that the browser always gets closed afterwards?

File: lighthouse-extension/test/extension-test.js (hunk 16 above)
Comment (wardpeet) [reply lighthouse-extension/test/extension-test.js]: no real reason :) whatever you guys prefer

File: lighthouse-extension/test/extension-test.js (hunk 18 above)
Comment (wardpeet) [reply lighthouse-extension/test/extension-test.js]: we could also to the lighthouse run in before and do regular test cases 

```
describe('lighthouse extension' () => {
it ('should contain all categories', () => {

});

it('should contain all audits of category performance', () => {
})
```

File: lighthouse-extension/test/extension-test.js (hunk 14 above)
Comment (wardpeet) [reply lighthouse-extension/test/extension-test.js]: filmstrip doesn't have a group either so I'll lose that one too. I could test for that separately.`

```
{id: 'screenshot-thumbnails', weight: 0},
{id: 'network-requests', weight: 0},
```

Comment (paulirish) [reply lighthouse-extension/test/extension-test.js]: good catch. IMO let's remove filmstrip from the selector above and use this `.length`.

File: lighthouse-extension/test/extension-test.js (hunk 16 above)
Comment (paulirish) [reply lighthouse-extension/test/extension-test.js]: arrow fn style makes sense if its a 1liner or 1.5 liner. past that I agree it should get a function declaration.

File: lighthouse-extension/test/extension-test.js (hunk 18 above)
Comment (paulirish) [reply lighthouse-extension/test/extension-test.js]: 👍 let's do that!

File: lighthouse-extension/test/extension-test.js (hunk 16 above)
Comment (wardpeet) [reply lighthouse-extension/test/extension-test.js]: let me memorize this rule for the next time :p

File: lighthouse-core/scripts/run-mocha.sh (hunk 15 above)
Comment (wardpeet) [reply lighthouse-core/scripts/run-mocha.sh]: something as `"unit-extension": "cd lighthouse-extension && yarn test"`?

Comment (paulirish) [reply lighthouse-core/scripts/run-mocha.sh]: bingo. so remove it from run-mocha. add `test-extension` to root manifest. add `test` to extension manifest. put `test-extension` in travis.yml

File: lighthouse-extension/test/extension-test.js [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+
+      const elementCount = await getAuditElementsCount({category, selector});
+
+      assert.equal(expected, elementCount,
+        `${category} does not have the correct amount of audits`);
+    }
+  });
+
+  it('should contain a filmstrip', async () => {
+    const filmstrip = await extensionPage.$('lh-filmstrip');
```
[the remark this answers is not in the document]
Comment (paulirish) [reply lighthouse-extension/test/extension-test.js]: i think this sorts it out:

```js
  it('should contain a filmstrip', async () => {
    const filmstrip = await extensionPage.$('.lh-filmstrip');
    assert.equal(true, !!filmstrip, `filmstrip is not available`);
  });
```

File: lighthouse-extension/test/extension-test.js [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+
+    assert.equal(null, filmstrip,
+      `filmstrip is not available`);
+  });
+
+  it('should contain no failed audits', async () => {
+    const auditErrors = await extensionPage.$$('.lh-debug');
+    const failedAudits = auditErrors.filter(error =>
```
Comment (paulirish) [inline lighthouse-extension/test/extension-test.js]: heh. I ran into a similar API misunderstanding. I was doing `$$eval(selector, el => ...`) but its actually `elems` instead of just `el`. Whoopsie daisy.

-----------

try something like this out:


```js
  it('should not have any audit errors', async () => {
    function getDebugStrings(elems) {
      return elems.map(el => ({
        debugString: el.textContent,
        title: el.closest('.lh-audit').querySelector('.lh-score__title').textContent,
      }));
    }

    const auditErrors = await extensionPage.$$eval('.lh-debug', getDebugStrings);
    const errors = auditErrors.filter(item => item.debugString.includes('Audit error:'));
    assert.deepStrictEqual(errors, [], 'Audit errors found within the report');
  });
```

gives a nice purdy error: 
image

**Comment by paulirish** [comment]
🎉 this is epic, @wardpeet. thx for making this happen. :D
